File: scripts/convert_to_flat_format.py
#!/usr/bin/env python3
"""
Convert semantic_output_data.json from nested JSON to flat key-value format.
Adds a 'flat_output' field to each training example.
"""

import json
import sys
from pathlib import Path


def flatten_semantic_output(semantic_output):
    """
    Convert nested semantic output to flat key:value format.
    
    Example:
    {
      "operation": "Create",
      "target_resource": {
        "type": "Webserver",
        "name": "jill",
        "properties": {"port": "8080", "runtime": "go"}
      },
      "context": {"user_role": "admin"}
    }
    
    Becomes:
    "operation:Create type:Webserver name:jill port:8080 runtime:go user_role:admin"
    """
    parts = []
    
    # Add operation
    if "operation" in semantic_output:
        parts.append(f"operation:{semantic_output['operation']}")
    
    # Add target_resource fields
    if "target_resource" in semantic_output:
        resource = semantic_output["target_resource"]
        
        if "type" in resource:
            # Simplify type (remove :: separators)
            resource_type = resource["type"].replace("::", "_")
            parts.append(f"type:{resource_type}")
        
        if "name" in resource:
            parts.append(f"name:{resource['name']}")
        
        # Add properties
        if "properties" in resource:
            for key, value in resource["properties"].items():
                parts.append(f"{key}:{value}")
    
    # Add context fields
    if "context" in semantic_output:
        for key, value in semantic_output["context"].items():
            parts.append(f"{key}:{value}")
    
    return " ".join(parts)


def convert_file(input_path, output_path):
    """Convert training data file to include flat_output field."""
    print(f"Reading {input_path}...")
    with open(input_path, 'r') as f:
        data = json.load(f)
    
    print(f"Converting {len(data)} examples...")
    converted = 0
    for example in data:
        if "semantic_output" in example:
            try:
                example["flat_output"] = flatten_semantic_output(example["semantic_output"])
                converted += 1
            except Exception as e:
                print(f"Warning: Failed to convert example: {example.get('query', 'unknown')}")
                print(f"  Error: {e}")
    
    print(f"Successfully converted {converted}/{len(data)} examples")
    
    print(f"Writing to {output_path}...")
    with open(output_path, 'w') as f:
        json.dump(data, f, indent=2)
    
    print("Done!")


def main():
    # Paths
    script_dir = Path(__file__).parent
    project_dir = script_dir.parent
    input_file = project_dir / "trainingdata" / "semantic_output_data.json"
    output_file = project_dir / "trainingdata" / "semantic_output_data_flat.json"
    
    if not input_file.exists():
        print(f"Error: Input file not found: {input_file}")
        sys.exit(1)
    
    convert_file(input_file, output_file)
    
    # Show example
    print("\n--- Example conversion ---")
    with open(output_file, 'r') as f:
        data = json.load(f)
    
    if data:
        example = data[0]
        print(f"Query: {example['query']}")
        print(f"Flat output: {example.get('flat_output', 'N/A')}")


if __name__ == "__main__":
    main()

File: scripts/convert_to_flat_format.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

// Using map[string]interface{} for flexibility, to mirror the Python script's dynamic nature.
type Example map[string]interface{}

func flattenSemanticOutput(semanticOutput map[string]interface{}) (string, error) {
	var parts []string

	// Add operation
	if op, ok := semanticOutput["operation"].(string); ok {
		parts = append(parts, "operation:"+op)
	}

	// Add target_resource fields
	if resource, ok := semanticOutput["target_resource"].(map[string]interface{}); ok {
		if resType, ok := resource["type"].(string); ok {
			// Simplify type (remove :: separators)
			resType = strings.ReplaceAll(resType, "::", "_")
			parts = append(parts, "type:"+resType)
		}
		if name, ok := resource["name"].(string); ok {
			parts = append(parts, "name:"+name)
		}

		// Add properties
		if properties, ok := resource["properties"].(map[string]interface{}); ok {
			for key, value := range properties {
				parts = append(parts, fmt.Sprintf("%s:%v", key, value))
			}
		}
	}

	// Add context fields
	if context, ok := semanticOutput["context"].(map[string]interface{}); ok {
		for key, value := range context {
			parts = append(parts, fmt.Sprintf("%s:%v", key, value))
		}
	}

	return strings.Join(parts, " "), nil
}

func convertFile(inputPath, outputPath string) {
	fmt.Printf("Reading %s...\n", inputPath)
	inputFile, err := ioutil.ReadFile(inputPath)
	if err != nil {
		fmt.Printf("Error: Failed to read input file: %v\n", err)
		os.Exit(1)
	}

	var data []Example
	if err := json.Unmarshal(inputFile, &data); err != nil {
		fmt.Printf("Error: Failed to parse JSON: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Converting %d examples...\n", len(data))
	converted := 0
	for _, example := range data {
		if semanticOutput, ok := example["semantic_output"].(map[string]interface{}); ok {
			flatOutput, err := flattenSemanticOutput(semanticOutput)
			if err != nil {
				query, _ := example["query"].(string)
				fmt.Printf("Warning: Failed to convert example: %s\n", query)
				fmt.Printf("  Error: %v\n", err)
				continue
			}
			example["flat_output"] = flatOutput
			converted++
		}
	}

	fmt.Printf("Successfully converted %d/%d examples\n", converted, len(data))

	fmt.Printf("Writing to %s...\n", outputPath)
	outputFile, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		fmt.Printf("Error: Failed to marshal output JSON: %v\n", err)
		os.Exit(1)
	}

	if err := ioutil.WriteFile(outputPath, outputFile, 0644); err != nil {
		fmt.Printf("Error: Failed to write output file: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Done!")

	// Show example
	fmt.Println("\n--- Example conversion ---")
	if len(data) > 0 {
		example := data[0]
		query, _ := example["query"].(string)
		flatOutput, _ := example["flat_output"].(string)
		fmt.Printf("Query: %s\n", query)
		fmt.Printf("Flat output: %s\n", flatOutput)
	}
}

func main() {
	// Get the directory of the executable to build the paths relative to the project root.
	// This mimics the behavior of the Python script's Path(__file__).parent.
	// For a `go run` command, this is less straightforward. We will assume the script
	// is run from the project root.
	projectDir, err := os.Getwd()
	if err != nil {
		fmt.Printf("Error getting current directory: %v\n", err)
		os.Exit(1)
	}

	inputFile := filepath.Join(projectDir, "trainingdata", "semantic_output_data.json")
	outputFile := filepath.Join(projectDir, "trainingdata", "semantic_output_data_flat.json")

	if _, err := os.Stat(inputFile); os.IsNotExist(err) {
		fmt.Printf("Error: Input file not found: %s\n", inputFile)
		os.Exit(1)
	}

	convertFile(inputFile, outputFile)
}
